File: app/src/main/java/com/example/jbmarks/dashboard/data/DashboardRepository.kt
package com.example.jbmarks.dashboard.data

import android.content.Context
import com.example.jbmarks.calendar.data.CalendarRepository
import com.example.jbmarks.chat.data.ChatRepository
import com.example.jbmarks.tasks.data.TasksRepository
import com.example.jbmarks.user.data.UserRepository
import java.text.SimpleDateFormat
import java.util.*

/**
 * Repository for dashboard data
 */
class DashboardRepository(context: Context) {
    
    private val userRepository = UserRepository(context)
    private val tasksRepository = TasksRepository(context)
    private val chatRepository = ChatRepository(context)
    private val calendarRepository = CalendarRepository()
    
    /**
     * Fetch all dashboard statistics
     */
    suspend fun getDashboardStats(): Result<DashboardStats> {
        return try {
            android.util.Log.d("DashboardRepository", "Starting to fetch dashboard stats...")
            
            // Fetch user info
            val userName = userRepository.getCurrentUser().getOrNull()?.fullName ?: "User"
            android.util.Log.d("DashboardRepository", "User: $userName")
            
            // Fetch tasks
            android.util.Log.d("DashboardRepository", "Fetching tasks...")
            val tasks = try {
                tasksRepository.getTasks()
            } catch (e: Exception) {
                android.util.Log.e("DashboardRepository", "Exception fetching tasks", e)
                e.printStackTrace()
                emptyList()
            }
            android.util.Log.d("DashboardRepository", "Fetched ${tasks.size} tasks")
            
            // Log task statuses for debugging
            tasks.forEach { task ->
                android.util.Log.d("DashboardRepository", "Task: ${task.id} - ${task.title} - Status: ${task.status}")
            }
            
            val activeTasks = tasks.count { it.status != com.example.jbmarks.tasks.domain.TaskStatus.COMPLETED }
            android.util.Log.d("DashboardRepository", "Active tasks: $activeTasks (out of ${tasks.size} total)")
            
            // Get recent active tasks (limit to 3 for display)
            val recentActiveTasks = tasks
                .filter { it.status != com.example.jbmarks.tasks.domain.TaskStatus.COMPLETED }
                .take(3)
            android.util.Log.d("DashboardRepository", "Recent active tasks: ${recentActiveTasks.size}, first task: ${recentActiveTasks.firstOrNull()?.title}")
            
            // Count completed tasks
            val completedToday = tasks.count { task ->
                task.status == com.example.jbmarks.tasks.domain.TaskStatus.COMPLETED
            }
            android.util.Log.d("DashboardRepository", "Completed tasks: $completedToday (out of ${tasks.size} total)")
            
            // Get recent completed tasks (limit to 3 for display)
            val recentCompletedTasks = tasks
                .filter { it.status == com.example.jbmarks.tasks.domain.TaskStatus.COMPLETED }
                .take(3)
            android.util.Log.d("DashboardRepository", "Recent completed tasks: ${recentCompletedTasks.size}, first task: ${recentCompletedTasks.firstOrNull()?.title}")
            
            // Fetch unread messages (from chat conversations)
            android.util.Log.d("DashboardRepository", "Fetching chats...")
            val chats = chatRepository.getRecentChats()
            android.util.Log.d("DashboardRepository", "Fetched ${chats.size} chats")
            
            // Count unread messages from all chats
            val unreadMessages = chats.sumOf { it.unreadCount }
            android.util.Log.d("DashboardRepository", "Unread messages: $unreadMessages")
            
            // Fetch upcoming events (don't fail if this fails)
            val upcomingEvents = try {
                android.util.Log.d("DashboardRepository", "Fetching events...")
                val events = calendarRepository.getCalendarEvents()
                val count = events.size
                android.util.Log.d("DashboardRepository", "Upcoming events: $count")
                count
            } catch (e: Exception) {
                android.util.Log.w("DashboardRepository", "Failed to fetch calendar events, using 0", e)
                0
            }
            
            val stats = DashboardStats(
                activeTasks = activeTasks,
                completedToday = completedToday,
                unreadMessages = unreadMessages,
                upcomingEvents = upcomingEvents,
                userName = userName,
                recentActiveTasks = recentActiveTasks,
                recentCompletedTasks = recentCompletedTasks
            )
            
            android.util.Log.d("DashboardRepository", "Final stats: activeTasks=${stats.activeTasks}, completedToday=${stats.completedToday}, unreadMessages=${stats.unreadMessages}, upcomingEvents=${stats.upcomingEvents}")
            
            Result.success(stats)
        } catch (e: Exception) {
            android.util.Log.e("DashboardRepository", "Failed to fetch dashboard stats", e)
            e.printStackTrace()
            Result.failure(e)
        }
    }
}

File: JbmrksIOs/JbmrksIOs/Services/DashboardRepository.swift
//
//  DashboardRepository.swift
//  JbmrksIOs
//
//  Repository for dashboard data aggregation
//

import Foundation

protocol DashboardRepository {
    func getDashboardStats() async throws -> DashboardStats
}

nonisolated class DashboardRepositoryImpl: DashboardRepository {
    private let tasksRepository: TasksRepository?
    private let chatRepository: ChatRepository?
    private let calendarRepository: CalendarRepository?
    private let activityFeedRepository: ActivityFeedRepository?
    private let userRepository: UserRepository?
    
    init(
        tasksRepository: TasksRepository?,
        chatRepository: ChatRepository?,
        calendarRepository: CalendarRepository?,
        activityFeedRepository: ActivityFeedRepository?,
        userRepository: UserRepository?
    ) {
        self.tasksRepository = tasksRepository
        self.chatRepository = chatRepository
        self.calendarRepository = calendarRepository
        self.activityFeedRepository = activityFeedRepository
        self.userRepository = userRepository
    }
    
    func getDashboardStats() async throws -> DashboardStats {
        // Fetch user info
        let userName = (try? await userRepository?.getCurrentUser().fullName) ?? "User"
        
        // Fetch tasks
        var tasks: [Task] = []
        if let tasksRepo = tasksRepository {
            tasks = try await tasksRepo.getTasks(responsibleId: nil, createdBy: nil, status: nil, groupId: nil)
        }
        
        let activeTasks = tasks.filter { $0.status != .completed }.count
        let completedTasks = tasks.filter { $0.status == .completed }
        let completedToday = completedTasks.count // Simplified - Android checks date
        
        // Get recent tasks
        let recentActiveTasks = Array(tasks.filter { $0.status != .completed }.prefix(3))
        let recentCompletedTasks = Array(completedTasks.prefix(3))
        
        // Fetch unread messages
        var unreadMessages = 0
        if let chatRepo = chatRepository {
            let chats = (try? await chatRepo.getRecentChats()) ?? []
            unreadMessages = chats.reduce(0) { $0 + $1.unreadCount }
        }
        
        // Fetch upcoming events
        var upcomingEvents = 0
        if let calendarRepo = calendarRepository {
            let events = (try? await calendarRepo.getCalendarEvents()) ?? []
            upcomingEvents = events.count
        }
        
        return DashboardStats(
            activeTasks: activeTasks,
            completedToday: completedToday,
            unreadMessages: unreadMessages,
            upcomingEvents: upcomingEvents,
            userName: userName,
            recentActiveTasks: recentActiveTasks,
            recentCompletedTasks: recentCompletedTasks
        )
    }
}
